src/provider.rs

```diff
@@ -903,95 +893,135 @@ fn compute_outboard(
+        CollectionBuilder {
+            sources: Vec::with_capacity(cap),
+        }
+    }
 
-    // insert outboards into the database and build collection
-    for (path, name, hash, outboard) in outboards {
-        debug_assert!(outboard.len() >= 8, "outboard must at least contain size");
-        let size = u64::from_le_bytes(outboard[..8].try_into().unwrap());
-        db.insert(
-            hash,
-            BlobOrCollection::Blob(Data {
-                outboard: Bytes::from(outboard),
-                path: path.clone(),
-                size,
-            }),
-        );
-        total_blobs_size += size;
-        // if the given name is `None`, use the filename from the given path as the name
-        let name = name.unwrap_or_else(|| {
-            path.file_name()
-                .and_then(|s| s.to_str())
-                .unwrap_or_default()
-                .to_string()
-        });
-        blobs.push(Blob { name, hash });
+    /// Add a file by its path.
+    pub fn add_file(&mut self, file: PathBuf) -> &mut Self {
+        self.add_source(DataSource::File(file))
     }
 
-    let c = Collection {
-        name: "collection".to_string(),
-        blobs,
-        total_blobs_size,
-    };
+    /// Add a file by its path, specifying a specific name.
+    pub fn add_named_file(&mut self, name: String, path: PathBuf) -> &mut Self {
+        self.add_source(DataSource::NamedFile { name, path })
+    }
 
-    let data = postcard::to_stdvec(&c).context("blob encoding")?;
-    let (outboard, hash) = abao::encode::outboard(&data);
-    let hash = Hash::from(hash);
-    db.insert(
-        hash,
-        BlobOrCollection::Collection((Bytes::from(outboard), Bytes::from(data.to_vec()))),
-    );
+    /// Convenience method for creating a collection from a list of files.
+    pub fn from_files(files: Vec<PathBuf>) -> Self {
+        let mut builder = Self::with_capacity(files.len());
+        for file in files {
+            builder.add_file(file);
+        }
+        builder
+    }
+
+    fn add_source(&mut self, source: DataSource) -> &mut Self {
+        self.sources.push(source);
+        self
+    }
 
-    Ok((db, entries, hash))
+    /// Create the actual collection.
+    pub async fn build(self) -> Result<(Database, Hash)> {
+        let (db, _, hash) = self.build_with_entries().await?;
+        Ok((db, hash))
+    }
+
+    /// Create the actual collection, including `ProvideResponseEntry`s.
+    /// Writing the data into the provided database.
+    async fn build_into_with_entries(
+        self,
+        db: &mut HashMap<Hash, BlobOrCollection>,
+    ) -> Result<(Vec<ProvideResponseEntry>, Hash)> {
+        // +1 is for the collection itself
+        let mut blobs = Vec::with_capacity(self.sources.len());
+        let mut total_blobs_size: u64 = 0;
+
+        // compute outboards in parallel, using tokio's blocking thread pool
+        let outboards = self.sources.into_iter().map(|data| {
```

Comment:
Not related to the builder refactor, but this code can fail if you open too many files concurrently.

I changed this into a stream and then used `.buffer_unordered(num_cpus.get())` in some branch to fix this.